enable ui-table to be configured via ui_control

Christian-Me opened the pull request:
<!--
## Before you hit that Submit button....

Please read our [contribution guidelines](https://github.com/node-red/node-red-nodes/blob/master/CONTRIBUTING.md)
before submitting a pull-request.

## Types of changes

What types of changes does your code introduce?
Put an `x` in the boxes that apply
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

<!--
If you want to raise a pull-request with a new feature, or a refactoring
of existing code, it **may well get rejected** if it hasn't been discussed on
the [forum](https://discourse.nodered.org) or
[slack team](https://nodered.org/slack) first.

-->

## Proposed changes

I'm not an expert in custom  ui nodes so sorry if I have done everything/something wrong. I`m not a professional programmer and neither are my javascript skills the best.

For my project it would be nice if the ui-table could be configured via msg.ui_control. All data is already in my flow so retyping everything into the node configuration in many instances was not a nice option. Just when I was looking for infos in the forum about ui-table I found out that I`m not the only one so I thought it was a good idea to do a pr.

So here it is, very simple implementation. If you send together with the msg.payload an array in msg.ui_control with one object per column like this it works perfectly. The property names do not match the names in the ui but this is what I found when i export an a configured table note. Here is my example:

![image](https://user-images.githubusercontent.com/15779507/67726141-2845c980-f9e5-11e9-842e-5817738ebe8a.png)

I do not know what the "formatterParams" object is good for and if it is used. If it is necessary but in all cases the same perhaps I should add it in the beforeEmit function. And perhaps "translate" the property names to match the names in the ui (formatter = format and field = property)

Sorry if I took the wrong path here. I did not dig very deep into your code.
Looking forward to your comments

## Checklist
<!-- Put an `x` in the boxes that apply -->

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red-nodes/blob/master/CONTRIBUTING.md)
- [ ] For non-bugfix PRs, I have discussed this change on the mailing list/slack team.
- [ ] I have run `grunt` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

Christian-Me commented:
Now I know what "formatterParams" is good for ... cool ....

![image](https://user-images.githubusercontent.com/15779507/67776631-dfcbf180-fa60-11e9-8156-8db92db1dd79.png)

I have to implement the ability to  pass functions to format values (_19.3°C_ instead _19.3_). or define color ranges.

Christian-Me commented:
Last but not least I added the ability to pass _color_ or _legend_ functions.

![image](https://user-images.githubusercontent.com/15779507/67812196-d06b9900-fa9e-11e9-83e8-3823fe65f006.png)

`msg.ui_control.formatterParams.legend="{return value + '°C';}"`

I know that there can be a potential security risk using the `Function("value","{return value + '°C';}")` constructor but I don't know how high / risky it is. Perhaps you have am opinion. When I understand it right all code entered in the Node-RED editor is somehow "build in" the runtime over the network.
Alternatively I can add a "fixed" logic to add a suffix defined by a ui.control parameter. 
But this would exclude the example in the 2nd column where I tested a little bit of logic operation `{return (value>0)? value+' %' : '-';}` Not tested color functions but this could be nice too.

Tell me what you think in principal about that feature.

and this would be the code in line 119 of node.js (could be optimized but work):
```
msg.ui_control.forEach(function (element) {
    if (element.hasOwnProperty("formatterParams")) {
        for (var prop in element.formatterParams) {
            if ((element.formatterParams.hasOwnProperty("color") ||
                    element.formatterParams.hasOwnProperty("legend")) &&
                    (typeof element.formatterParams[prop] === "string") && 
                    element.formatterParams[prop].includes('{') && 
                    element.formatterParams[prop].includes('}')) {
                
                element.formatterParams[prop]= Function("value",element.formatterParams[prop].slice(element.formatterParams[prop].indexOf('{')+1,element.formatterParams[prop].lastIndexOf('}')));
            }
        }
    }
});
```

dceejay commented on node-red-node-ui-table/node.js:
```diff
@@ -64,7 +64,10 @@ module.exports = function (RED) {
                     forwardInputMessages: false,
                     beforeEmit: function (msg, value) {
-                        return { msg: { payload: value } };
+                        if (msg.hasOwnProperty("ui_control") && Array.isArray(msg.ui_control))
```
all ui_control messages should be objects not arrays - so you can add multiple properties in on message

Christian-Me replied:
Hi, thank you for your reply. You are right, Just done that and added `ui_control.layout` and `ui_control.movableColumns` to the now called `ui_control.columns` object.
All looks OK ... but I recognized one problem. When I switch tabs the column sizes are reduced to minimum. All other parameters of the columns object are still working only the width information somehow now working. Spending a few hours with the debugger I still found no reason because the columns attributes are still there and passed to the tabulator... Works on reload page and initial start but not when I come form another tab.
Any idea? Or any information why create table is sometimes called twice (first with no tabledata and no columns and next correct. And there is this setInterval thing I do not know why.
Sorry that that pr is a little bit "unfinished". Thought it whould be a easy little addition.

Christian-Me replied:
This is was happens after switching tabs:
![image](https://user-images.githubusercontent.com/15779507/67901112-40912200-fb66-11e9-9fa7-2495f7884ee7.png)
and this after reload or sending a message to the node
![image](https://user-images.githubusercontent.com/15779507/67901176-70d8c080-fb66-11e9-858d-c1ed800bb016.png)
Perhaps I have to reinstall the original version and see if this happens there too.

Christian-Me replied:
Hi, again. The "original" current version shows the same problem. So my modification seams not to cause that problem.
An I also learned that the Interval / document.querySelector(tablediv) is there to prevent this or other issues caused by calling createTable when the container/div(?) is not ready. In my case this did not prevent the tabulator to get the wrong width (240 something instead of the correct  850 something) (I`m using chrome Version 78.0.3904.70 on windows).
Raising the timer to a 400 (instead of 40) solve the problem (for me) but seams not to be a good solution.
I'm not into HTML, CSS and so on (my knowledge if any is from the mid 90th) so I will need little help here. I can live with the flaw (either raising the timeout or sending data to the node) but  for now I have to leave this issue to someone more capable than me, sorry.

Christian-Me commented:
200ms is workling here ... is is NOT hapening after page refresh it happens after changing tabs. Thats what makes me wonder if there is another node-red mechanism changing the width later in the progress. Perhaps during placing the groups. But I have no experience what is happening. During searching around I updated tabulator to the latest version. OK? 

Just added grouping and group headers and feedbacks for column sorting and width changes to be able to make sorting and layout changes persistent. What is better, a second output for these data:

![image](https://user-images.githubusercontent.com/15779507/68030907-c8884080-fcba-11e9-888e-8abad7f7f848.png)

![image](https://user-images.githubusercontent.com/15779507/68030953-e190f180-fcba-11e9-944f-7e39f88bd54e.png)

or sending a msg.ui_control object over the standard output? This can break nodes not expecting messages with no toppic / mayload but keeps it lesss complicated. Version 1 is done but i can switch back to version 2. The second output only appear when you thick the "send on layout change". Will add group open/close events too.

It could look like this if you send a `groupBy `field and a `groupHeader `function

![image](https://user-images.githubusercontent.com/15779507/68031315-b5c23b80-fcbb-11e9-947a-5f47e32e15ce.png)

Working to long on this but its already quite usefull for my projects and perhaps for others too.
(when/if ready to merge done I`m willing to update the help file too)

dceejay commented:
Hi - I was using a hand fixed version of tabulator as there were these issues I found that do now seem to be closed - just need to check they have been.
https://github.com/olifolkerd/tabulator/issues?utf8=%E2%9C%93&q=is%3Aissue++dceejay+

Maybe try 100mS then if 200 working :-)

What is the second output for - please do NOT add it to this PR - we can make that another PR if necessary.

Christian-Me commented:
Resizing & rearranging columns is fun but useless for me if I don't get feedback. So my question was "how do you like the feedback to be given: as an extra output, or as an extra object like msg.ui_control (or not at all ;). In most cases there will be a new message with no valid topic/payload when i.e. a column was resized or a row was collapsed. This could perhaps break some flows.

I can take the feedback part out now, because I had already do a "workaround" with the msg.origin object to block messages bounced back to the frontend when emitted ... Have no idea why and how to avoid it. The original plan was, if I can store the new arranged column layout I can avoid the problem with the column/table width not updating correctly when i use percentages. The resized columns are send back by pixel.
Or as a compromise: Send the ui_control updates through the standard output if the checkbox is ticked (default off).

I can go back to the original version of tabulator as it did not resolved the issue.

dceejay commented:
I'm not saying no - just do it as a separate PR so I can digest it more easily and think about it.
Ideally all changes should be discussed via the discourse forum first so the wider community can join in and add other input, as there may be better ways - or other features we didn't think of.
Certainly I would want the original author @kazuhitoyokoi and @hotNipi to have a look.

Christian-Me commented:
What started as a few lines of code ended in a bigger thing than I expected. I promise I do it next time better, I apologize. And I understand you perfectly no worries. I appreciate your work, time and trouble you invest in this project. 
I'm still no used to GitHub and instead of creating a new branch after a PR, I continue implementing all my ideas still in my head, sorry.
Let me finished the things we already discussed (I removed the 2nd output and the checkbox in the configuration UI, because it is not needed at all (if you pass a callback function you should be aware that you will get callbacks) Removed the upgrade of tabulator.
I will cleanup my code and do some further test over the weekend.  

Perhaps it is the best to delete this PR and start with a proper discussion and do a new PR?

dceejay commented:
Hi - ok - yes keep them clean...
This PR - add ui_control inputs
Another PR - replace the library with latest
Third PR - Add extra output with feedback of width changes etc.
thanks

Christian-Me commented:
Hi, it took longer than expected or I could not spend as much time as expected on this,
![image](https://user-images.githubusercontent.com/15779507/68468823-1a7c1980-0219-11ea-91d7-da0bdd9eb4af.png)

How it works now.
- The `msg.ui_control.tabulator` object will be (deep) cloned. If in the process a function is found the `Function` constructor is used. Other objects inside ui_control are copied (see example of `ui_control.customHeight`)
- When the table is created the data in `$scope.config.ui_control` is used. If one of the basic configs (layout, columns, autoColumns and movableColumns) is missing defaults are used to maintain the same behavior as the before.
- Callbacks: I removed the callbacks for columnResized and columnMoved. However it is possible to pass these (and others) by sending them via `ui_control.tabulator`. To give the callbacks the ability to send messages back the `$scope` is bound to functions defined through ui_control. The Callback should use `this.send()` instead of `$scope.send()`

I started to  write [readme.md](https://github.com/Christian-Me/node-red-contrib-home/tree/master/Node-RED/dashboard/ui-table) which could be added to this nodes documentation.

Open Issue: I realized that messages send by `$scope.send()` or `this.send()` are received by `$scope.$watch('msg', function (msg)`. This not only triggers an unnecessary rebuild of the table it also can cause problems with callbacks. I could not find a way to avoid this behavior. **Perhaps somebody can help here.**. The workaround is if a message contains `ui_control.callback` the $watch function is ended right in the beginning (see Line 124).

Looking forward to hear from you.

Christian-Me commented:
Hi, I know we are all very busy and I appreciate all the hard work you are contributing to this awesome project. Perhaps you have forgotten my last commit or even missed it. I don’t want to put  any pressure on this because it works fine on my side. I think perhaps others can take advantage too ;)
Just tell me if there is anything I can do or you expect from my side.
Regards Chris.

dceejay commented:
Ideally I'd like @kazuhitoyokoi  to have a look also as it was originally his work.

Christian-Me commented:
That’s fine and totally understood. 😃

kazuhitoyokoi commented:
Thank you for your consideration. I'll check it tomorrow (within 24 hours).

kazuhitoyokoi commented:
It works great! I learned how to use `msg.ui_control`. 🥇

dceejay commented:
Thanks @kazuhitoyokoi & @Christian-Me

dceejay commented:
I have merged this.
@Christian-Me - I think it would be good to add a "simple" example of the extra functionality to the examples folder if possible ?

Christian-Me commented:
Great to hear that you like it.
Next I plan to make it possible to add/delete/update lines or cells without sending the hole table. I plan to do it with ui_control messages to keep the topic/payload as simple as it is.
I like to use ui-table for a syslog server for all my esp8266 nodes and expect a substantial amount of data to be transferred. The existing filter functions in tabulator seems ideal for this task. (And it can be a reusable demo how to take advantage of ui_control messages)

Christian-Me commented:
Will prepare a simple example. Bargraph with values and colors and groups should be ok?
